integrations/jenkins/client.py

```diff
@@ -45,6 +46,36 @@ async def get_builds(self) -> AsyncGenerator[list[dict[str, Any]], None]:
             event.attributes.setdefault(ResourceKey.BUILDS, []).extend(builds)
             yield builds
 
+    async def get_build_stages(self, build_url: str) -> list[dict[str, Any]]:
+        response = await self.client.get(f"{build_url}/wfapi/describe")
+        response.raise_for_status()
+        stages = response.json()["stages"]
+
+        # Enrich stages with full URLs
+        for stage in stages:
+            relative_url = stage["_links"]["self"]["href"]
+            stage["__fullUrl"] = f"{self.jenkins_base_url}{relative_url}"
+
+        return stages
+
+    async def get_stages(self) -> AsyncGenerator[list[dict[str, Any]], None]:
+        if cache := event.attributes.get(ResourceKey.STAGES):
+            logger.info("picking jenkins stages from cache")
+            yield cache
+            return
+
+        async for builds in self.get_builds():
+            stages: list[dict[str, Any]] = []
+            for build in builds:
+                build_url = build["url"]
+                try:
+                    stages.extend(await self.get_build_stages(build_url))
+                except Exception as e:
+                    logger.error(f"Failed to get stages for build {build_url}: {e}")
```

Comment:
if the builds are not independent, we can try to run these in parallel using async gather